wiki/mapping/map-format.md

```diff
@@ -313,16 +313,456 @@ The exact specifics of what goes in `_customData` is entirely dependent on commu
+
+### Rotation Events
+
+```json
+{
+  "b": 10.0, // Beat
+  "e": 0,    // Event type
+  "r": 15.0, // Rotation
+}
+```
+
+#### b
+The time, in beats, where this object reaches the player.
+
+#### e
+|`e`|Result|
+|:---:|-----|
+|`0`|Early rotation. Rotates future objects, while also rotating objects at the same time.|
+|`1`|Late rotation. Rotates future objects, but ignores rotating objects at the same time.|
+
+#### r
+A float which represents clockwise rotation.
+
+### Color Notes
+
+```json
+{
+  "b": 10.0, // Beat
+  "x": 1,
+  "y": 0,
+  "c": 0,    // Color
+  "d": 1,    // Direction
+  "a": 0     // Angle offset
+}
+```
+
+#### b
+The time, in beats, where this object reaches the player.
+
+#### x
+An integer number, from 0 to 3, which represents the column where this note is located.
+The far left column is located at index 0, and increases to the far right column located at index 3.
+
+#### y
+An integer number, from 0 to 2, which represents the layer where this note is located.
+The bottommost layer is located at layer 0, and inceases to the topmost layer located at index 2.
+
+#### c
+An integer which represents the color of the note.
+|`_c`|Result|
+|:---:|-----|
+|`0`|Red|
+|`1`|Blue|
+
+#### d
+This indicates the cut direction for the note.
+
+|`_d`|Result|
+|:-------------------:|-------------------|
+|`0`|Up|
+|`1`|Down|
+|`2`|Left|
+|`3`|Right|
+|`4`|Up Left|
+|`5`|Up Right|
+|`6`|Down Left|
+|`7`|Down Right|
+|`8`|Any (Dot Note)|
+
+#### a
+A integer number which represents the number of additional counter-clockwise angle offset of note
+
+### Bomb Notes
+
+```json
+{
+  "b": 10.0, // Beat
+  "x": 1,
+  "y": 0
+}
+```
+
+#### b
+The time, in beats, where this object reaches the player.
+
+#### x
+An integer number, from 0 to 3, which represents the column where this note is located.
+The far left column is located at index 0, and increases to the far right column located at index 3.
+
+#### y
+An integer number, from 0 to 2, which represents the layer where this note is located.
+The bottommost layer is located at layer 0, and inceases to the topmost layer located at index 2.
+
+### Obstacles
+
+```json
+{
+  "b": 10.0, // Beat
+  "x": 1,
+  "y": 0,
+  "d": 5,    // Duration
+  "w": 1,    // Width
+  "h": 3     // Height
+}
+```
+
+#### b
+The time, in beats, where this object reaches the player.
+
+#### x
+An integer number, from 0 to 3, which represents the column where this obstacle is located.
+The far left column is located at index 0, and increases to the far right column located at index 3.
+
+#### y
+An integer number, from 0 to 2, which represents the layer where base of the obstacle is located.
+The bottommost layer is located at layer 0, and inceases to the topmost layer located at index 2.
+|`y`|Result|
+|:-------------------:|-------------------|
+|`1`|Grounded obstacle|
+|`2`|Prone obstacle|
+|`3`|Crouch obstacle|
```

Comment:
Text description says 0 to 2, but table is 1 to 3.

Reply:
Table do be wrong.